search/search_query.cpp

```diff
@@ -1609,44 +1615,99 @@ void Query::SearchLocality(MwmValue const * pMwm, Locality & res1, Region & res2
 
   auto codingParams = trie::GetCodingParams(pMwm->GetHeader().GetDefCodingParams());
 
-  ModelReaderPtr searchReader = pMwm->m_cont.GetReader(SEARCH_INDEX_FILE_TAG);
-
-  using TValue = FeatureIndexValue;
-  auto const trieRoot = trie::ReadTrie<SubReaderWrapper<Reader>, ValueList<TValue>>(
-      SubReaderWrapper<Reader>(searchReader.GetPtr()), SingleValueSerializer<TValue>(codingParams));
-
-  ForEachLangPrefix(params, *trieRoot, [&](TrieRootPrefix & langRoot, int8_t lang)
+  auto versionFormat = pMwm->GetHeader().GetFormat();
+  if (versionFormat < version::v7)
   {
-    impl::DoFindLocality doFind(*this, pMwm, lang);
-    MatchTokensInTrie(params.m_tokens, langRoot, doFind);
+    using TValue = FeatureWithRankAndCenter;
+
+    ModelReaderPtr searchReader = pMwm->m_cont.GetReader(SEARCH_INDEX_FILE_TAG);
+
+    auto const trieRoot = trie::ReadTrie<SubReaderWrapper<Reader>, ValueList<TValue>>(
+        SubReaderWrapper<Reader>(searchReader.GetPtr()),
+        SingleValueSerializer<TValue>(codingParams));
+
+    ForEachLangPrefix(params, *trieRoot, [&](TrieRootPrefix<TValue> & langRoot, int8_t lang)
+                      {
+                        impl::DoFindLocality doFind(*this, pMwm, lang);
```

Comment:
Please, create a separate template class or a template function, as this code is duplicated below.